data_import: Check for invalid user emails when converting data.

PieterCK opened the pull request:
This PR makes the third-party data converters check for invalid user emails. If it finds any, it’ll raise an Exception and show an error message with all the bad emails listed out.

<details>

<summary>Error log when there are invalid user emails</summary>

```

Converting data ...
2024-10-10 05:48:17.320 INFO [] ######### IMPORTING USERS STARTED #########

2024-10-10 05:48:17.349 INFO [] USLACKBOT: slackbot -> imported-slackbot-bot@invalid.c
2024-10-10 05:48:17.351 INFO [] U06NU4E26M9: pieterceka123 -> pieterceka123@gmail.com
2024-10-10 05:48:17.363 INFO [] U07NFSVQCAD: pabbly_connect -> pabblyconnect-bot@invalid.c
Traceback (most recent call last):
  File "/srv/zulip/./manage.py", line 150, in <module>
    execute_from_command_line(sys.argv)
  File "/srv/zulip/./manage.py", line 115, in execute_from_command_line
    utility.execute()
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/__init__.py", line 436, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 413, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/srv/zulip/zerver/lib/management.py", line 97, in execute
    super().execute(*args, **options)
  File "/srv/zulip-venv-cache/b477df627afb33e3ce1cd594561fcbb9d3946a14/zulip-py3-venv/lib/python3.10/site-packages/django/core/management/base.py", line 459, in execute
    output = self.handle(*args, **options)
  File "/srv/zulip/zerver/management/commands/convert_slack_data.py", line 79, in handle
    do_convert_zipfile(
  File "/srv/zulip/zerver/data_import/slack.py", line 1416, in do_convert_zipfile
    do_convert_directory(slack_data_dir, output_dir, token, threads, convert_slack_threads)
  File "/srv/zulip/zerver/data_import/slack.py", line 1460, in do_convert_directory
    ) = slack_workspace_to_realm(
  File "/srv/zulip/zerver/data_import/slack.py", line 182, in slack_workspace_to_realm
    ) = users_to_zerver_userprofile(slack_data_dir, user_list, realm_id, int(NOW), domain_name)
  File "/srv/zulip/zerver/data_import/slack.py", line 362, in users_to_zerver_userprofile
    validate_user_emails_for_import(list(found_emails))
  File "/srv/zulip/zerver/data_import/import_util.py", line 845, in validate_user_emails_for_import
    raise ValidationError(error_log)
django.core.exceptions.ValidationError: ['Invalid email format, please fix the following email(s) and try again: imported-slackbot-bot@invalid.c, export-bot-bot@invalid.c, pabblyconnect-bot@invalid.c']

```
</details>
Fixes: #31783


**Screenshots and screen captures:**

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented on zerver/data_import/mattermost.py:
```diff
@@ -132,15 +132,11 @@ def convert_user_data(
+        if check_user_in_team(raw_item, team_name) or raw_item["is_mirror_dummy"]:
+            user = process_user(raw_item, realm_id, team_name, user_id_mapper)
+            user_handler.add_user(user)
+    user_handler.check_user_emails()
```
Other than adding `user_handler.check_user_emails()`, I also did a little refactor here.

PieterCK commented on zerver/data_import/import_util.py:
```diff
@@ -825,3 +827,21 @@ def long_term_idle_helper(
     return long_term_idle
+
+
+def check_converted_users_email(user_emails: list[str]) -> None:
```
This function is here and not in `user_handler.py` inside `check_user_emails()` because unlike Mattermost and Rocket.chat importers, the Slack importer doesn't use the `UserHandler` class.

laurynmm replied:
Maybe `check_imported_user_emails`? Or `validate_imported_user_emails`?

laurynmm replied:
Or `validate_user_emails_for_import`?

PieterCK replied:
hmm, I guess `validate_user_emails_for_import` is ok. I avoided using the word "imported" because at this point we're still converting the Slack export.

PieterCK commented on zerver/tests/test_slack_importer.py:
```diff
@@ -374,7 +374,7 @@ def test_users_to_zerver_userprofile(self, mock_get_data_file: mock.Mock) -> Non
             "Xf023DSCdd": {"value": "employer"},
         }
-        user_data = [
+        user_data: list[dict[str, Any]] = [
```
type hint is added to this variable because Mypy was complaining when I tampered with the users' emails for the new tests just a few lines below.

PieterCK commented:
I wonder if we should just use `SystemExit` instead of `Exception` because we don't really need the full traceback for when bad emails are detected.

alya commented:
@laurynmm do you have the context for reviewing this one?

laurynmm commented on zerver/data_import/import_util.py:
```diff
@@ -825,3 +827,21 @@ def long_term_idle_helper(
+            f"Invalid email format, please fix the following email(s) and try again:\n\n{details}"
+        )
+        # ValidationError isn't used here because it's error message
+        # is formatted as literal string, so it won't output newlines(\n).
```
Could we just print a string with the invalid emails separated by commas? I do think getting all the invalid emails at once is a good choice, but I don't expect there to be an extremely high number of them most of the time.

`Invalid email format, please fix the following email(s) and try again: user1@email.c, user4@email`

PieterCK replied:
Yes, that works too. I'll switch back to raising `ValidationError` too because we don't use newlines. I've updated the error log in the description

laurynmm commented on zerver/data_import/mattermost.py:
```diff
@@ -132,15 +132,11 @@ def convert_user_data(
-    for raw_item in user_data_list:
-        user = process_user(raw_item, realm_id, team_name, user_id_mapper)
-        user_handler.add_user(user)
+    for raw_item in user_data_map.values():
```
This would be easier to read with `user` or `user_data` instead of `raw_item`.

PieterCK replied:
switched to `user_data`.

laurynmm reviewed (commented):
@PieterCK - Thanks for adding this email validation for imports! I had a few comments, but overall it looks good. Let me know if you have any questions about my feedback.

It might make sense for @prakhar1144 to do the next round of review since he's worked a bit more with the import code than I have.

PieterCK commented:
@laurynmm Thanks for reviewing! I've pushed changes to address your comments and updated the error log in the description. Let me know if I missed anything :+1:

laurynmm commented:
@PieterCK - The updates look good!

Do you think the function in the `UserHandler` class should be updated to `validate_user_emails` vs `check_user_emails`? Also, even though they are the same value at this point in the import process, it might be good to use "delivery_email" in that function since that's what Zulip considers to be the user's real email address once the `UserProfile` is created.

prakhar1144 commented:
I wan't able to take a look today. I'll review this tomorrow.

PieterCK commented:
> Do you think the function in the `UserHandler` class should be updated to `validate_user_emails` vs `check_user_emails`? Also, even though they are the same value at this point in the import process, it might be good to use "delivery_email" in that function since that's what Zulip considers to be the user's real email address once the `UserProfile` is created.

Updated:
- Switched `check_user_emails` to `validate_user_emails`.
- Used "delivery_email" in `validate_user_emails` instead of "email".

laurynmm commented:
@PieterCK - These updates look good to me at this point. So, even though @prakhar1144 hasn't had a chance to review yet, I'm going to mark this as ready for integration review.

timabbott commented:
This is great, merged, thanks @PieterCK! I marked this as a backport candidate, since we'll want to ship the fix in 9.x as well.

timabbott commented:
Backported in #32129.
